### PR TITLE
add policy validation call

### PR DESCRIPTION
**Pull request by ChrisSchreiber**
- Add attest policy method to protobuf
- Implement fetching and encoding resource occurrences

Method is still WIP

### CODE REVIEW

File: server/server.go [hunk 1]
```diff
@@ -32,6 +36,43 @@ func (r *rodeServer) BatchCreateOccurrences(ctx context.Context, occurrenceReque
...
 }
 
+func (r *rodeServer) AttestPolicy(ctx context.Context, request *pb.AttestPolicyRequest) (*pb.AttestPolicyResponse, error) {
+	log := r.logger.Named("AttestPolicy")
+	log.Debug("received requests")
+
+	// check OPA policy has been loaded
+
+	// fetch occurrences from grafeas
+	listOccurrencesResponse, err := r.grafeasClient.ListOccurrences(ctx, &grafeas.ListOccurrencesRequest{Filter: fmt.Sprintf("resourceUri:%s", request.ResourceURI)})
```
Comment (mrparkers) [inline server/server.go:46]: not sure if this filter string is correct.  thought it'd be something like `resource.uri = %s`

File: server/server.go [hunk 2]
```diff
@@ -32,6 +36,43 @@ func (r *rodeServer) BatchCreateOccurrences(ctx context.Context, occurrenceReque
...
+	if err != nil {
+		log.Error("failed to list occurrences for resource", zap.NamedError("error", err), zap.String("resource", request.ResourceURI))
+		return nil, err
+	}
+
+	// json encode occurrences
+	_, err = json.Marshal(listOccurrencesResponse)
```
Comment (mrparkers) [inline server/server.go:53]: you'll want to use the `jsonpb` package to do marshalling / unmarshalling of protobufs

File: server/server.go [hunk 3]
```diff
@@ -32,6 +36,43 @@ func (r *rodeServer) BatchCreateOccurrences(ctx context.Context, occurrenceReque
...
+func (r *rodeServer) AttestPolicy(ctx context.Context, request *pb.AttestPolicyRequest) (*pb.AttestPolicyResponse, error) {
+	log := r.logger.Named("AttestPolicy")
+	log.Debug("received requests")
+
+	// check OPA policy has been loaded
+
+	// fetch occurrences from grafeas
+	listOccurrencesResponse, err := r.grafeasClient.ListOccurrences(ctx, &grafeas.ListOccurrencesRequest{Filter: fmt.Sprintf("resourceUri:%s", request.ResourceURI)})
+	if err != nil {
+		log.Error("failed to list occurrences for resource", zap.NamedError("error", err), zap.String("resource", request.ResourceURI))
```
Comment (mrparkers) [inline server/server.go:48]: nitpick: you can just do `zap.Error(err)` instead of `zap.NamedError`

File: server/server_test.go [hunk 4]
```diff
@@ -14,45 +18,112 @@ import (
...
 			}
-		})
+			// mocked Grafeas BatchCreateOccurrences response
+			grafeasBatchCreateOccurrencesResponse := &grafeas_proto.BatchCreateOccurrencesResponse{
+				Occurrences: []*grafeas_proto.Occurrence{
+					randomOccurrence,
+				},
+			}
+			// ensure Grafeas BatchCreateOccurrences is called with expected request and inject response
+			grafeasClient.EXPECT().BatchCreateOccurrences(gomock.AssignableToTypeOf(context.Background()), gomock.Eq(grafeasBatchCreateOccurrencesRequest)).Return(grafeasBatchCreateOccurrencesResponse, nil)
```
Comment (mrparkers) [inline server/server_test.go:59]: is there a way to pre-configure this response in a `BeforeEach`?  we usually try to capture these setup steps there and leave the `It` for the actual assertions.

File: server/server_test.go [hunk 5]
```diff
@@ -14,45 +18,112 @@ import (
...
+			resourceURI := gofakeit.URL()
+			policy := gofakeit.Word()
+			listOccurrencesRequest := &grafeas_proto.ListOccurrencesRequest{
+				Filter: fmt.Sprintf("resourceUri:%s", resourceURI),
+			}
+			listOccurrencesResponse := &grafeas_proto.ListOccurrencesResponse{
+				Occurrences: []*grafeas_proto.Occurrence{
+					createRandomUnspecifiedOccurrence(),
+				},
+			}
```
Comment (mrparkers) [inline server/server_test.go:97]: I'd probably move some of this stuff up to a common `BeforeEach` as well

**Review by mrparkers — COMMENTED** [review]
nice work with `gomock`, it's cool that we can remove all of that test boilerplate code.

File: server/server_test.go (hunk 4 above)
Comment (ChrisSchreiber) [reply server/server_test.go:61]: I moved the initialization of the request and response objects into a `BeforeEach` block which makes sense. I'm not as sure about the mock itself. I don't think we can separate asserting the method was called from injecting the response so i think it makes sense to leave it in the `It` block but i'm curious what you think?

File: proto/v1alpha1/rode-attest.proto [hunk 6]
```diff
@@ -0,0 +1,31 @@
+syntax = "proto3";
+
+package rode.v1alpha1;
+
+option go_package = "github.com/rode/rode/proto/v1alpha1";
+
+import "google/protobuf/timestamp.proto";
+
+message AttestPolicyRequest {
+  string policy = 1;
```
Comment (aalsabag) [inline proto/v1alpha1/rode-attest.proto:10]: should these all be required fields? I assume this message would be useless without a policy passed in

**Review by aalsabag — APPROVED** [review]

File: proto/v1alpha1/rode-attest.proto (hunk 6 above)
Comment (ChrisSchreiber) [reply proto/v1alpha1/rode-attest.proto:10]: Good call. I'll include that in the next PR.

**Pull request merged**